### PR TITLE
Add MaxRL for Policy gradients

### PR DESCRIPTION
**Pull request by Athe-kunal**

### CODE REVIEW

**Comment by Athe-kunal** [comment]
@natolambert and @zafstojano 

This PR is for adding MaxRL: https://arxiv.org/abs/2602.02710

Weights and Biases report: https://wandb.ai/athe_kunal/rlhf-book/runs/npnh0yno

Let me know about the PR

**Comment by natolambert** [comment]
@Athe-kunal I looked, generally looks right. I don't love how the abstraction is unfolding, but that's not necessarily a you problem. Probably can merge soon, maybe @zafstojano will look quickly soon.

**Comment by Athe-kunal** [comment]
What feedback do you have on the abstraction @natolambert?

**Comment by zafstojano** [comment]
Pushed a few simplifications on top:

- Moved `binary_reward = correctness * (format == 1.0)` onto `Experience` so it's computed once in the rollout. `compute_maxrl_advantages` no longer recomputes it, and the `wandb.log` call goes back to the simple string-only `avg` lambda (`avg("binary_reward")`).
- Threaded `cfg: Config` into `apply_reward_kl` and `compute_advantages` instead of unpacking individual fields, so both call sites in `rollout.py` collapse to one line each.
- Aligned `configs/maxrl.yaml` with the GRPO/DAPO defaults (`num_rollouts: 8`, `temperature: 0.6`, `data.size: 3000`) — `num_rollouts: 16` was OOMing on my hardware.
- Trimmed the duplicated baseline/advantage formula out of `MaxRLLoss`'s docstring since it lives in `compute_maxrl_advantages`.

(generated with Claude Code)

**Comment by zafstojano** [comment]
@Athe-kunal I would suggest for future PR's to create a new branch on your remote, and open a PR from that one instead of your fork's main. That way the PR wouldn't have the long history of commits at the beginning.

**Comment by Athe-kunal** [comment]
Sorry, my bad, I missed it, thanks @zafstojano

**Comment by zafstojano** [comment]
### Refactor: per-completion rewards collapsed into a `TensorDict`

Two follow-up commits on top of the MaxRL work consolidate the per-completion reward bookkeeping. Motivation: every new reward signal (binary, future ones) was adding a new field to `Experience`, a new positional arg to `compute_advantages` / `apply_reward_kl`, and new lines in the wandb logger and the console summary. With more reward components on the way that growth wasn't going to scale.

#### Change

- `Experience` now has a single `rewards: TensorDict` field instead of `rewards / correctness / format / penalties / binary_reward`. Keys: `total / correctness / format / penalty / binary`.
- `compute_rewards(...)` returns the `TensorDict` directly (including the `binary` derived signal — it no longer lives in `rollout.py`).
- `apply_reward_kl` and `compute_advantages` consume the `TensorDict`. The `binary_reward` arg on `compute_advantages` is gone.
- Wandb metrics and the rollout console summary both iterate `rewards.keys()` and emit `avg_<key>`. **Adding a new reward signal is now one line in `compute_rewards` — no other consumer needs to change.**
- Buffer collation: `split_experience_batch` uses `val.unbind(0)` (works for both `Tensor` and `TensorDict`); `join_experiences_batch` dispatches to `torch.stack(vals, dim=0)` for `TensorDict` entries and `pad_sequences` for variable-length tensor fields.
- Adds [`tensordict`](https://pypi.org/project/tensordict/) (`>=0.12.2`) as a dependency.

#### Compatibility

Wandb metric names changed: old `avg_reward / avg_correctness_reward / avg_format_reward / avg_response_penalty / avg_binary_reward` → new `avg_total / avg_correctness / avg_format / avg_penalty / avg_binary`. Historical runs are not comparable by name. Numeric semantics are unchanged.

#### `tensordict` reference

Docs: https://docs.pytorch.org/tensordict/stable/index.html

It's a small library from the PyTorch org that gives you a dict-of-tensors object with a declared `batch_size`. All entries agree on the leading shape; trailing dims, dtype, and device can differ per key. Standard tensor ops (`indexing`, `stack`, `unbind`, `.to(device)`, `.reshape`, `.clone`, `.cat`, `.apply`) propagate over every key — which is what makes the buffer collation simplify so cleanly.

**Comment by zafstojano** [comment]
@natolambert given the large changes, would recommend re-running all algorithms. 

Quick local run seems fine:
```
───────────────────────────────────────── 8/3000 (0.3%) ─────────────────────────────────────────
╭─────────────────────────────────────── Rollout Results ───────────────────────────────────────╮
│ total: 0.9364    correctness: 0.5653    format: 0.7422    penalty: 0.0000    binary: 0.5653   │
╰───────────────────────────────────────────────────────────────────────────────────────────────╯
╭─────────────────────────────────────────── Sample ────────────────────────────────────────────╮
│  Question:     Spell this word backward (example: sun -> nus): ingleside                      │
│  Oracle:       ediselgni                                                                      │
│  Completion:   <think>                                                                        │
│                Okay, let's see. The user wants me to spell "ingleside" backward. The example  │
│                given was "sun" becoming "nus". So, the process is straightforward: take the   │
│                word and reverse the order of the letters.                                     │
│                                                                                               │
│                First, I need to write out the word "ingleside" and then reverse it. Let me    │
│                check the letters. The word is i-n-g-l-e-s-i-d-e. Wait, let me count again.    │
│                Let's break it down:                                                           │
│                                                                                               │
│                i (1), n (2), g (3), l (4), e (5), s (6), i (7), d (8), e (9). So nine         │
│                letters total. Reversing them would start from the last letter to the first.   │
│                So the reversed word would be e-d-i-s-e-l-g-n-i. Let me check that again.      │
│                Original: 1 2 3 4 5 6 7 8 9 → reversed is 9 8 7 6 5 4 3 2 1. So the letters    │
│                in reverse order are e, d, i, s, e, l, g, n, i. Putting them together:         │
│                ediselgn i? Wait, let me write it step by step.                                │
│                                                                                               │
│                Original: i n g l e s i d e                                                    │
│                                                                                               │
│                Reversed: e d i s e l g n i                                                    │
│                                                                                               │
│                So combining those letters: e d i s e l g n i. Let me check if that's          │
│                correct. The original word is "ingleside". Let's reverse each character:       │
│                ... (truncated)                                                                │
╰───────────────────────────────────────────────────────────────────────────────────────────────╯

  Loss: -0.0135 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16/16 * 0:00:09
```

**Comment by Athe-kunal** [comment]
@zafstojano, the addition of TensorDict is really useful

Can we just put the entire Experience as TensorClass, too? It would simplify a lot of handling, and we can still do attribute type extraction, like experience.advantages. It can be handled in another PR

**Comment by zafstojano** [comment]
Let's keep the setup as such for now, and if we notice it is becoming a bottleneck we can  use it later on

**Comment by Athe-kunal** [comment]
Hi @natolambert and @zafstojano 
Let me know if I can help to get this PR merged on

Also, I recently made a YouTube video explaining RL algorithms. I would love to get your feedback on how I can improve, any inconsistencies, and how I can explain things better. Would really appreciate it

RL Algorithms: https://youtu.be/G9YKtuhwN5g?si=F_sRkBVUDiNpojpA

LLM lectures playlist: https://www.youtube.com/playlist?list=PLtYQQQmDwvJsAJ9Tc9iy5MEGpldvRip3V

**Comment by natolambert** [comment]
Am re-running the baseline locally, should be good to merge soon.
